src/SSTFrequencyLimitDetector.cpp

```diff
@@ -0,0 +1,209 @@
+                    avx2_freq = m_bucket_avx2_frequency[bucket_idx];
+                    avx512_freq = m_bucket_avx512_frequency[bucket_idx];
+                }
+
+                for (const auto core_idx : cores_in_package) {
+                    // Two neighboring bins in the SST-TF table might or might
+                    // not be equal, so check both.
+                    if (observed_core_frequencies[core_idx] > avx2_freq ||
+                        observed_core_frequencies[core_idx] >= sse_freq) {
+                        m_core_frequency_limits[core_idx] = m_sse_hp_tradeoffs;
+                        m_core_lp_frequencies[core_idx] = m_low_priority_sse_frequency;
+                    }
+                    else if (observed_core_frequencies[core_idx] > avx512_freq ||
+                             observed_core_frequencies[core_idx] >= avx2_freq) {
+                        m_core_frequency_limits[core_idx] = m_avx2_hp_tradeoffs;
+                        m_core_lp_frequencies[core_idx] = m_low_priority_avx2_frequency;
+                    }
+                    else {
+                        m_core_frequency_limits[core_idx] = m_avx512_hp_tradeoffs;
+                        m_core_lp_frequencies[core_idx] = m_low_priority_avx512_frequency;
+                    }
+                }
+            }
+            else {
+                // SST-TF is available, but disabled. Assume any core can reach
+                // the current max observed frequency across cores.
+                const auto core_with_max_frequency = *std::max_element(
+                    cores_in_package.begin(), cores_in_package.end(),
+                    [&observed_core_frequencies](int lhs, int rhs) {
+                        return observed_core_frequencies[lhs] < observed_core_frequencies[rhs];
+                    });
+                const auto max_frequency = observed_core_frequencies[core_with_max_frequency];
+                for (const auto core_idx : cores_in_package) {
+                    m_core_frequency_limits[core_idx] = {
+                        // Single-element vector because this is the only
+                        // tradeoff presented without SST-TF enabled.
+                        {cores_in_package.size(), max_frequency}};
+                    m_core_lp_frequencies[core_idx] = m_cpu_frequency_sticker;
+                }
+            }
+        }
+    }
+
+    std::vector<std::pair<unsigned int, double> >
+        SSTFrequencyLimitDetector::get_core_frequency_limits(unsigned int core_idx) const
+    {
+#ifdef GEOPM_DEBUG
+        if (core_idx >= m_core_frequency_limits.size()) {
+            throw Exception("SSTFrequencyLimitDetector::" + std::string(__func__) +
+                                "Encountered invalid core index: " +
+                                std::to_string(core_idx),
+                            GEOPM_ERROR_INVALID,
+                            __FILE__, __LINE__);
+        }
+#endif
```

Comment:
I'm a bit worried about this check being hidden behind a ``GEOPM_DEBUG`` "flag". I'm tempted to say this should be turned into an assertion for two reasons: (1) it seems like passing an invalid ``core_idx`` would be a programmer error, (2) the behavior of ``std::vector`` when passing an index out of bounds is undefined, which would result in an issue that's difficult to detect and debug.

If there are scenarios that could result in invalid input (i.e. not programmer error), I would say to simply go ahead and detect the situation and throw an exception.